tests/L0/run_optimizers/test_fused_novograd.py

```diff
@@ -0,0 +1,168 @@
+                    state['exp_avg_sq'] = torch.zeros([]).to(state['exp_avg'].device)
+                    if amsgrad:
+                        # Maintains max of all exp. moving avg. of sq. grad. values
+                        state['max_exp_avg_sq'] = torch.zeros([]).to(state['exp_avg'].device)
+
+                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
+                if amsgrad:
+                    max_exp_avg_sq = state['max_exp_avg_sq']
+                beta1, beta2 = group['betas']
+
+                state['step'] += 1
+
+                norm = torch.sum(torch.pow(grad, 2))
+
+                if exp_avg_sq == 0:
+                    exp_avg_sq.copy_(norm)
+                else:
+                    exp_avg_sq.mul_(beta2).add_(norm, alpha=1 - beta2)
+
+                if amsgrad:
+                    # Maintains the maximum of all 2nd moment running avg. till now
+                    torch.max(max_exp_avg_sq, exp_avg_sq, out=max_exp_avg_sq)
+                    # Use the max. for normalizing running avg. of gradient
+                    denom = max_exp_avg_sq.sqrt().add_(group['eps'])
+                else:
+                    denom = exp_avg_sq.sqrt().add_(group['eps'])
+
+                grad.div_(denom)
+                if group['weight_decay'] != 0:
+                    grad.add_(p.data, alpha=group['weight_decay'])
+                if group['grad_averaging']:
+                    grad.mul_(1 - beta1)
+                exp_avg.mul_(beta1).add_(grad)
+
+                p.data.add_(exp_avg, alpha=-group['lr'])
+        
+        return loss
+
+
+class TestFusedNovoGrad(TestFusedOptimizer):
+
+    def __init__(self, *args, **kwargs):
+        super(TestFusedNovoGrad, self).__init__(*args, **kwargs)
+
+        # The options for NovoGrad and FusedNovoGrad are very specific if they
+        # are expected to behave the same.
+        self.options = {'lr':1e-3, 'betas':(0.95, 0), 'eps':1e-8,
+                 'weight_decay':0, 'grad_averaging':False, 'amsgrad':False}
+        
+        self.tst_options = {'lr':1e-3, 'betas':(0.95, 0), 'eps':1e-8,
+                 'weight_decay':0, 'grad_averaging':False, 'amsgrad':False, 
+                 'bias_correction':False, 'reg_inside_moment':True, 
+                 'norm_type':2, 'init_zero':False, 'set_grad_none':True}
+
+        self.ref_optim = Novograd
+        self.fused_optim = apex.optimizers.FusedNovoGrad
+
+    def test_float(self):
+        self.gen_single_type_test(param_type=torch.float)
+
+    def test_half(self):
+        self.gen_single_type_test(param_type=torch.float16)
+
+    @unittest.skipIf(torch.cuda.device_count()<2, "more than 1 GPU required")
+    def test_multi_device(self):
+        devices = ("cuda:1", "cuda:0")
+        for current_dev, tensor_dev in product(devices, devices):
+            with torch.cuda.device(current_dev):
+                print("Some info from the test", current_dev, tensor_dev)
```

Comment:
Could you remove this `print` statement or is this needed for the test?

Reply:
I've removed it.